health-services/ExerciseSampleCompose/app/src/main/java/com/example/exercisesamplecompose/data/HealthServicesRepository.kt

```diff
@@ -0,0 +1,117 @@
+
+    private var exerciseService: ForegroundService? = null
+
+    suspend fun hasExerciseCapability() = getExerciseCapabilities() != null
+
+    private suspend fun getExerciseCapabilities() = exerciseClientManager.getExerciseCapabilities()
+
+    suspend fun isExerciseInProgress() = exerciseClientManager.isExerciseInProgress()
+
+    suspend fun isTrackingExerciseInAnotherApp() =
+        exerciseClientManager.isTrackingExerciseInAnotherApp()
+
+    fun prepareExercise() = exerciseService?.prepareExercise()
+    fun startExercise() = exerciseService?.startExercise()
+    fun pauseExercise() = exerciseService?.pauseExercise()
+    fun endExercise() = exerciseService?.endExercise()
+    fun resumeExercise() = exerciseService?.resumeExercise()
+
+    var bound = mutableStateOf(false)
+
+    var exerciseServiceState: MutableState<ServiceState> = mutableStateOf(ServiceState.Disconnected)
+
+    private val connection = object : android.content.ServiceConnection {
+        override fun onServiceConnected(className: ComponentName, service: IBinder) {
+            val binder = service as ForegroundService.LocalBinder
+            binder.getService().let {
+                exerciseService = it
+                exerciseServiceState.value = ServiceState.Connected(
+                    exerciseMetrics = it.exerciseMetrics,
+                    exerciseLaps = it.exerciseLaps,
+                    exerciseDurationUpdate = it.exerciseDurationUpdate,
+                    locationAvailabilityState = it.locationAvailabilityState,
+                    activeDurationUpdate = it.exerciseDurationUpdate.value,
+                    exerciseState = it.exerciseState,
+                    exerciseStateChange = it.exerciseStateChange,
+                )
+            }
+            bound.value = true
+        }
+
+        override fun onServiceDisconnected(arg0: ComponentName) {
+            bound.value = false
+            exerciseService = null
+            exerciseServiceState.value = ServiceState.Disconnected
+        }
+
+    }
+
+    fun createService() {
+        Intent(applicationContext, ForegroundService::class.java).also { intent ->
+            applicationContext.startService(intent)
+            applicationContext.bindService(intent, connection, Context.BIND_AUTO_CREATE)
+        }
+    }
+
+}
+
+/** Store exercise values in the service state. While the service is connected,
+ * the values will persist.**/
+sealed class ServiceState {
+    object Disconnected : ServiceState()
+    data class Connected(
+        val exerciseMetrics: StateFlow<DataPointContainer?>,
+        val exerciseLaps: StateFlow<Int>,
+        val exerciseDurationUpdate: StateFlow<ActiveDurationUpdate?>,
+        val locationAvailabilityState: StateFlow<LocationAvailability>,
+        val activeDurationUpdate: ActiveDurationUpdate?,
+        val exerciseState: StateFlow<ExerciseState>,
+        val exerciseStateChange: StateFlow<ExerciseStateChange>,
+    ) : ServiceState()
+}
```

Comment:
We should be able to "flatten" this class definition to

```kotlin
data class Connected(
        val exerciseMetrics: DataPointContainer?,
        val exerciseLaps: Int,
        val exerciseDurationUpdate: ActiveDurationUpdate?,
        val locationAvailabilityState: LocationAvailability,
        val activeDurationUpdate: ActiveDurationUpdate?,
        val exerciseState: ExerciseState,
        val exerciseStateChange: ExerciseStateChange,
    ) : ServiceState()
```

A `Flow` of `Flows` is seldom necessary, and has quite a bit of semantic overhead. We should be able to flatten it after the suggestion in the `Service`.

Reply:
Flattened!